### ISSUE TITLE
unconstrain argument to extract/as.array/as.matrix

### ISSUE THREAD

**Issue by jgabry** [stan-dev/rstan#648]
We really need an option to unconstrain the draws when using the standard ways to extract the draws from a stanfit object. 

I’m not sure this is possible to implement with the way unconstrain_pars() currently works, but it would make many things we want to implement so much easier (e.g. adaptive loo, bayesplot for unconstrained draws, etc.)

This is related to #630 and various other issues.

**Comment by bgoodri** [comment]
I think we just need the `unconstrain_pars` function to default to `pars = NULL` in order to return the unconstrained parameters for all iterations. I implemented this already and then the code got disappeared due to a bad merge of another PR.

**Comment by jgabry** [comment]
If I set `pars=NULL` when calling `unconstrain_pars()` I get an error: 

```r
fit <- stan_demo("eight_schools")
unconstrain_pars(fit, pars = NULL)
```
```
Error in object@.MISC$stan_fit_instance$unconstrain_pars(pars) : 
  Exception: Variable mu missing  (in 'model6edcd5e7762_eight_schools' at line 7)
```

Did I misunderstand your suggestion?

**Comment by bgoodri** [comment]
That is the right syntax but the code to implement it is gone.

**Comment by jgabry** [comment]
Oh I did misread what you wrote, sorry. 

So the code you wrote/modified is not hiding in the git history anywhere? Was it tricky to implement?

**Comment by bgoodri** [comment]
I don't think it was that tricky to implement but it is tricky to find.

**Comment by jgabry** [comment]
Ok so in the simplest case I think we can just do something like this:

```r
fit <- stan_demo("eight_schools")

# this code can easily be abstracted and go into the `unconstrain_pars()` method
tmp <- rstan:::create_skeleton(fit@model_pars, fit@par_dims)
skeleton <- tmp[names(tmp) != "lp__"]
constrained_draws <- as.matrix(fit)
S <- nrow(constrained_draws)
unconstrained_draws <- sapply(seq_len(S), FUN = function(s) {
  constrained_s <- rstan:::rstan_relist(constrained_draws[s, ], skeleton)
  fit@.MISC$stan_fit_instance$unconstrain_pars(constrained_s)
})
unconstrained_draws <- t(unconstrained_draws)
```

But if the constrained and unconstrained parameters don't have the same number of elements (e.g. for  `cholesky_factor_corr`, `simplex`, etc.) then `unconstrained_draws` will be very hard to interpret because `foo@.MISC$stan_fit_instance$unconstrain_pars` will just drop the extra dimensions and it will be hard for the user to figure out the correspondence between the columns of the unconstrained and constrained draws. So even though the code above works in all cases (I think) I don't think it's releasable until we figure out some logic for either 

(a) having informative column names for the unconstrained parameters, or 
(b) adding back in the missing columns that get dropped. 

Here's an example of a Stan program for which the output from `unconstrain_pars()` (if implemented using my method above) would be hard to interpret.  `as.matrix()` would return a matrix with 210 columns (not including lp__) whereas `unconstrain_pars()` would return a matrix with like `55` columns or something like that, and no column names:

```stan
parameters {
  cholesky_factor_corr[10] L_Omega;
  vector<lower=0>[10] tau;
}
transformed parameters {
  matrix[10, 10] L_Sigma = diag_pre_multiply(tau, L_Omega);
}
model {
  tau ~ normal(0, 1);
  L_Omega ~ lkj_corr_cholesky(1);
}
```

**Comment by bob-carpenter** [comment]
For simplex, it's just K - 1 parameters.  For correlation matrix, it's (N choose 2) arranged below the diagonal in a matrix, and for covariance matrix, it's (N choose 2) + N arranged in a lower-triangular matrix.  For unit vector, you can't back out the unconstrained given the constrained as they're non-deterministic, but you can choose the representative identity map.

Or you can just label them from 1 to whatever, which is what we do in the doc for the reference manual.

**Comment by jgabry** [comment]
Thanks @bob-carpenter. I think the problem we have in using that information is that RStan (as far as I know) has no way of identifying which parameter names correspond to which underlying parameter _types_, at least not without us doing some tricky parsing of the Stan program in R. In other words, we don't have a lookup table or other metadata coming from Stan that lets us check if `L_Omega` was declared as `cholesky_factor_corr`.

**Comment by bob-carpenter** [comment]
Right.  We were going to add that to the generated model code, but the stanc3 project has put all improvements to the parser on hold until we recreate the existing C++ functionality in OCaml.  Then hopefully these things will be easier to add.  But it's going to be months before we'll get there, at least.
